proposals/p2188.md

```diff
@@ -0,0 +1,664 @@
+}
+```
+
+The scrutinee is required to be of struct type, and to have the same set of
+field names as the pattern. The pattern is matched left-to-right, meaning that
+matching is performed in the field order specified in the pattern, not in the
+field order of the scrutinee. This is consistent with the behavior of matching
+against a struct-valued expression, where the expression pattern becomes the
+left operand of the `==` and so determines the order in which `==` comparisons
+for fields are performed.
+
+In the case where a field will be bound to an identifier with the same name, a
+shorthand syntax is available: `a: T` is synonymous with `.a = a: T`.
+
+```
+match ({.a = 1, .b = 2}) {
+  case {a: i32, b: i32} => { return a + b; }
+}
+```
+
+If some fields should be ignored when matching, a trailing `, _` can be added to
+specify this:
+
+```
+match ({.a = 1, .b = 2}) {
+  case {.a = 1, _} => { return 1; }
+  case {b: i32, _} => { return b; }
+}
+```
+
+This is valid even if all fields are actually named in the pattern.
+
+### Alternative patterns
+
+An alternative pattern is used to match one alternative of a choice type.
+
+-   _proper-pattern_ ::= _callee-expression_ _tuple-pattern_
+-   _proper-pattern_ ::= _designator_ _tuple-pattern_?
+
+Here, _callee-expression_ is syntactically an expression that is valid as the
+callee in a function call expression, and an alternative pattern is
+syntactically a function call expression whose argument list contains at least
+one _proper-pattern_.
+
+If a _callee-expression_ is provided, it is required to name a choice type
+alternative that has a parameter list, and the scrutinee is implicitly converted
+to that choice type. Otherwise, the scrutinee is required to be of some choice
+type, and the designator is looked up in that type and is required to name an
+alternative with a parameter list if and only if a _tuple-pattern_ is specified.
+
+The pattern matches if the active alternative in the scrutinee is the specified
+alternative, and the arguments of the alternative match the given tuple pattern
+(if any).
+
+```
+choice Optional(T:! Type) {
+  None,
+  Some(T)
+}
+
+match (Optional(i32).None) {
+  // ✅ `.None` resolved to `Optional(i32).None`.
+  case .None => {}
+  // ✅ `.Some` resolved to `Optional(i32).Some`.
+  case .Some(n: i32) => { Print("{0}", n); }
+  // ❌ Error, no such alternative exists.
+  case .Other => {}
+}
+
+class X {
+  external impl as ImplicitAs(Optional(i32));
+}
+
+match ({} as X) {
+  // ✅ OK, but expression pattern.
+  case Optional(i32).None => {}
+  // ✅ OK, implicitly converts to `Optional(i32)`.
+  case Optional(i32).Some(n: i32) => { Print("{0}", n); }
+}
+```
+
+Note that a pattern of the form `Optional(T).None` is an expression pattern and
+is compared using `==`.
+
+### Templates
+
+If the type of the scrutinee of a pattern is either a template parameter or is
+an associated type that depends on a template parameter, any checking of the
```

Comment:
What about other expressions dependent on a template parameter like `template T:! type, v: Vector(T)`?

Reply:
My intent was that we would be able to reject a tuple or struct pattern that tries to match `v` when checking the template definition. Hm, maybe that doesn't play well with the phases model for templates: if we don't symbolically evaluate `Vector(T)` because it involves a template parameter, then I guess we can't really ask whether it's a tuple or struct type.

Reply:
Maybe the definition of dependent from #2200 could be used here?

Reply:
I'm not sure that helps. For example, if we have `a: Vector(T)` and `b: Identity(T)`, I would expect both to have the same properties -- both type expressions are known to be of type `Type`, and for both of them the actual type value depends on a template parameter -- but for `a` we might know something about how to do pattern matching whereas for `b` we do not.

Also, given the extensibility support in https://github.com/carbon-language/carbon-lang/pull/2187, I suppose it might be the case that even for `a` we don't actually know how to do matching until we know `T` -- the `impl as Match` that's used might depend on `T`.